### ISSUE TITLE
pickle memoization bug with numpy

### ISSUE THREAD

**Issue by bmabey** [joblib/joblib#393]
Pickle memoization causes incorrect value-based hashes of dictionaries (and probably any other datastructure) containing numpy values.  This snippet illustrates the problem in a reduced case:

```python
import joblib as jl
import numpy as np

num = np.int64(10)

with_reuse = {'key0': num, 'key1': num}
without_reuse = {'key0': num, 'key1': np.int64(10)}

assert with_reuse == without_reuse
assert jl.hash(with_reuse) == jl.hash(without_reuse)
```
```
--------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-1-becd01458d06> in <module>()
      8 
      9 assert with_reuse == without_reuse
---> 10 assert jl.hash(with_reuse) == jl.hash(without_reuse)

AssertionError: 
```
[This notebook](
https://gist.github.com/bmabey/c4ddfa0bd5ac0db70a735c230d61c48c) shows that the problem lies with the memoization of the numpy values.

To fix this problem it seems like memoization will need to be disabled for numpy datatypes similar to [what is being done for strings currently](https://github.com/joblib/joblib/blob/master/joblib/hashing.py#L96-L103).

If you agree with this proposed fix I can create a PR for it.

**Comment by GaelVaroquaux** [comment]
> num = np.int64(10)

> with_reuse = {'key0': num, 'key1': num}
> without_reuse = {'key0': num, 'key1': np.int64(10)}

> assert with_reuse == without_reuse
> assert jl.hash(with_reuse) == jl.hash(without_reuse)

That's the correct behavior: those two objects are different, hence they
must hash to different values. To give you a more provocative example:

<pre>
l = list()
a = [l, l]
b = [l, list()]
</pre>

a and b must hash to different values. Indeed, a[0].append(1) will have a
very different behavior from b[0].append(1).

**Comment by bmabey** [comment]
When the objects are mutable, as in my original case, I can understand that logic...

For immutable objects I don't think that should be the case though but it seems to be:

```python
import joblib as jl
import numpy as np

num = np.int64(10)

with_reuse = (num, num)
without_reuse = (num, np.int64(10))

assert with_reuse == without_reuse
assert jl.hash(with_reuse) == jl.hash(without_reuse) # AssertionError
```

**Comment by GaelVaroquaux** [comment]
How can we know if an object is mutable or not? A tuple  (immutable) can contain a mutable. 

We are just following Python's logic. It's also the best way to ensure that joblib.Memory does not change the behavior of a code base.

**Comment by bmabey** [comment]
"We are just following Python's logic."

I see that and I agree with the reasoning. In actual my use case I have value objects that I know will not be mutated. So, in that setting it would be useful to optionally turn off memoization so the hashing is on the value level. Would you be open to that kind of option making it into the library?

**Comment by bmabey** [comment]
I see a way I can disable memoization for my use case by subclassing your `Pickler`. So, if you don't think having an option to disable memoization would be generally useful then you can close this issue.

**Comment by lesteve** [comment]
You are right, the issue seems to be the Pickler memoization (in one case pickle use the reference and in one case the value). Maybe we should think harder about which types pickle memoization should be disabled for.

I'll leave this one open for now.

**Comment by lesteve** [comment]
@GaelVaroquaux just to follow-up on the discussion after lunch. In the numpy array case, memoization is actually disabled as I said. Here is a snippet:

```py
import joblib
import numpy as np

arr = np.array([2., 3., 1., 4.])
print(joblib.hash((arr, arr)))
print(joblib.hash((arr, arr.copy())))
```

Output:
```
5780116027d93477ba792355f8261746
5780116027d93477ba792355f8261746
```

**Comment by GaelVaroquaux** [comment]
> import joblib
> import numpy as np

> arr = np.array([2., 3., 1., 4.])
> print(joblib.hash((arr, arr)))
> print(joblib.hash((arr, arr.copy())))

> Output:

> 5780116027d93477ba792355f8261746
> 5780116027d93477ba792355f8261746

That's a horrible bug that will lead someone to shoot himself in the foot
at some point.

**Comment by lesteve** [comment]
> That's a horrible bug that will lead someone to shoot himself in the foot at some point.

To put this in context for further reference (@GaelVaroquaux is that what you had in mind?):

```py
import numpy as np

from joblib import Memory

mem = Memory('/tmp/test')


def test(array_2_tuple):
    arr1, arr2 = array_2_tuple
    arr1[:] = 1
    return arr2.mean()

cached = mem.cache(test)
arr = np.zeros(10)
print('test(arr, arr):', test((arr, arr)))
arr = np.zeros(10)
print('cached(arr, arr):', cached((arr, arr)))

arr = np.zeros(10)
print('test(arr, arr.copy()):', test((arr, arr.copy())))
arr = np.zeros(10)
print('cached(arr, arr.copy()):', cached((arr, arr.copy())))
```

Output:
```
test(arr, arr): 1.0
________________________________________________________________________________
[Memory] Calling __main__--tmp-test_joblib_hash.test...
test(( array([ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.]),
  array([ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.])))
_____________________________________________________________test - 0.0s, 0.0min
cached(arr, arr): 1.0
test(arr, arr.copy()): 0.0
cached(arr, arr.copy()): 1.0
```

I would argue that joblib is not meant to cache non-pure functions so if you are trying to do that you kind are already looking for trouble ...

**Comment by GaelVaroquaux** [comment]
> I would argue that joblib is not meant to cache non-pure functions so if you
> are trying to do that you kind are already looking for trouble ...

Agreed. But we should minimize user suprise, in particular for users that
know Python well.

**Comment by bmabey** [comment]
As a user, with admittedly relatively little Python experience, I find the current behavior surprising that results in extra computation and storage since all of the functions I'm caching are pure. That said, you are probably right that this surprise is far less confusing than if someone ran into the something like the example just posted.

One potential option would be to allow users to denote that a function is pure or not. If it is it can use a hash that is based on the values alone while ignoring the reference differences. Otherwise the default behavior would be the conservative existing approach.
